Allow backorder product availability

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

As part of #738.

This PR adds a `backorder` availability for the products when back-ordering is enabled for a product and its stock status is on backorder (less than or equal to zero).

This also includes some refactoring, mainly applied to the admin inputs and attributes. A new set of input classes are created that each represents the input for an attribute. This way, the attribute's label, and description are defined in their input, which seems more logical given that they're only used in the Admin and not anywhere else the attributes are used.

A new DateTime input type is introduced that uses a set of two HTML inputs (`input type="date"` and `input type="time"`) to get the availability date and time from the user. I first tried to use a third-party DateTime picker library for this but decided to go native HTML instead because these inputs are supported by most modern browsers and by using them we avoid adding yet another library (albeit small) to our package.

This DateTime input assumes that the time entered by the user is based on the site's local timezone and not the browser's; therefore, the time is always converted to UTC when stored in the database and then back to the site's timezone when displaying to the user.

### Compatibility with [WooCommerce Pre-Orders](https://woocommerce.com/products/woocommerce-pre-orders/)

Note that since this PR was the base for the Pre-Order PR (https://github.com/woocommerce/google-listings-and-ads/pull/959) and the other one got merged, any code related to Pre-Orders is previously reviewed and can be safely ignored. Although it would be a good idea to test these two functionalities together. 

For example, since the pre-orders integration hooks into the `woocommerce_gla_product_attribute_values` action, it will override the availability date set by the user in the GLA section if pre-orders is enabled for the product and an availability date value is set in the Pre-Orders extension.

### Detailed test instructions:

1. Create a new product or edit an existing one
2. From `Inventory` enable `Manage Stock?` and `Allow Backorders?` for that product
3. Make sure it has a negative or zero stock quantity
4. Under the GLA tab of the product edit page, set the availability date to a valid date. You can also set the time but it's optional. Time is set in the site's timezone but stored in DB in UTC, so please check that the timezone conversion applies correctly.
5. Update the product and wait for it to be synced to Merchant Center
6. Confirm that the product's availability in MC is now set to "Backorder" instead of "In Stock" and the availability date and time are set correctly (also check the timezone).
7. Remove the availability date and update the product
8. Confirm that the availability date is deleted from product metadata and MC

**Test compatibility with WooCommerce Pre-Orders**

1. Install and activate [WooCommerce Pre-Orders](https://woocommerce.com/products/woocommerce-pre-orders/)
2. Enable pre-orders for the product and set an availability date via the Pre-Orders extension
3. Set a different availability date in the GLA section and also allow back orders for the product
4. Update the product and wait for it to sync
5. Confirm that the product availability is now set to "preorder" and its availability date is the one set via the Pre-Orders extension


### Changelog entry

> Add - Allow backorder stock availability for products

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted. 
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Or leave the "Changelog entry" header in place without any summary if no changelog entry is needed.  
Otherwise, the title of Pull Request will be used as the changelog entry.  
-->

nima-karimi committed:
Allow backorder product availability

If backorder is allowed for a WooCommerce product, and it's currently on `backorder` set its MC availability to 'backorder'.

nima-karimi committed:
Add unit tests for backorder availability

mikkamp commented:
As mentioned in #738 if we set the availability to `backorder` it's required to set the [`availability_date`](https://support.google.com/merchants/answer/6324470). I'm not seeing an option to set that in the PR. Is that going to prevent the product from being approved or is it no longer the case?

nima-karimi commented:
That's definitely an issue. Thanks for catching that. I thought that was only required for "preorders" but seems like it's also required for backorders. I'll make this PR WIP and try to add it here.

nima-karimi committed:
Add integration with WooCommerce Pre Orders extension

This will also add support for the `preorder` product availability.

nima-karimi committed:
Add validation constraint for availability date

nima-karimi committed:
Merge branch 'feature/add-backorder' into feature/add-preorder-integration

nima-karimi committed:
Add validation constraint for availability date

nima-karimi committed:
Fix doc comment

nima-karimi committed:
Get UTC avail date from meta value instead of extension class

nima-karimi committed:
Trigger an update job if pre-order disabled for product

nima-karimi committed:
Update src/Product/ProductHelper.php

Co-authored-by: Mik <mikkamp@users.noreply.github.com>

nima-karimi committed:
Add extra comments for `trigger_sync`

nima-karimi committed:
Merge pull request #959 from woocommerce/feature/add-preorder-integration

Add integration with WooCommerce Pre Orders extension

nima-karimi committed:
Add method to WP proxy to get timezone

nima-karimi committed:
Add DateTimeUtility service

nima-karimi committed:
Use WP proxy to get site timezone

nima-karimi committed:
Add DateTime admin input

nima-karimi committed:
Create `AttributeInputInterface`

nima-karimi committed:
Create separate input classes for attributes

These inputs are extended classes from the admin inputs and each represent an input for an attribute. Their label and description is set separately on their constructor based on the attribute type.

nima-karimi committed:
Use attribute default input instead of guessing it

nima-karimi committed:
Add AvailabilityDate attribute

nima-karimi committed:
Make sure to only update attributes once for each product

nima-karimi commented:
I've updated the code and the PR description. This is ready for review.

mikkamp commented on views/inputs/datetime.php:
```diff
@@ -0,0 +1,54 @@
+ */
+$input = $this->input;
+
+$input['class']         = $input['class'] ?? '';
+$input['placeholder']   = $input['placeholder'] ?? '';
+$input['style']         = $input['style'] ?? '';
+$input['wrapper_class'] = $input['wrapper_class'] ?? '';
+$input['name']          = $input['name'] ?? $input['id'];
+$input['desc_tip']      = $input['desc_tip'] ?? false;
+$input['date']          = $input['date'] ?? '';
+$input['time']          = $input['time'] ?? '';
+
+// Custom attribute handling
+$custom_attributes = [];
+if ( ! empty( $input['custom_attributes'] ) && is_array( $input['custom_attributes'] ) ) {
+	foreach ( $input['custom_attributes'] as $attribute => $value ) {
+		$custom_attributes[] = esc_attr( $attribute ) . '="' . esc_attr( $value ) . '"';
+	}
+}
```
Why do we add so many of these which we don't even use? Do we allow filtering input fields?

nima-karimi replied:
I aimed to keep it similar to other WooCommerce inputs so that if we need to customize it further, we could easily do so. But I guess we could also add these customizations later if we ever need them so I've removed most of them in https://github.com/woocommerce/google-listings-and-ads/pull/951/commits/49b7409d19b8fc91688851fa66948ff1aceca0c5

mikkamp reviewed (approved):
I was initially expecting the availability date to appear where we set the backorder status. Would it be helpful to add a link on that tab to GLA indicating availability date must be set?

Another question I had was whether we need any restrictions on whether the date is in the future? And what happens once the date is reached do we continue syncing with the old availability date?

Regarding the refactor for input and attributes, I'd say name and description could be useful in other locations, like exposing the attributes within the API. It just seems like more work to maintain an attribute by having to add/edit two classes.

I mainly just had questions / comments about the code, but none of those are blockers. When testing I get the results as expected. So I'll go ahead and mark this as approved.

nima-karimi committed:
Remove unnecessary input variables

nima-karimi committed:
Apply a better style to the channel visibility box

nima-karimi commented:
> Would it be helpful to add a link on that tab to GLA indicating availability date must be set?

I think this could be useful but there are some parameters that should be taken into account like whether stock management is enabled or listen on the JS events for the stock status change to display or hide this text. To keep things simple, I didn't add those here, but definitely a good idea for further improvements. Added an issue to follow up: https://github.com/woocommerce/google-listings-and-ads/issues/978

> Another question I had was whether we need any restrictions on whether the date is in the future?

I don't think that's something we need to worry about. In my opinion we just need to provide the input and leave the validation to Google.

> And what happens once the date is reached do we continue syncing with the old availability date?

Yes, we do. This is also something that I think we should leave to the user. If, for example, we add the extra functionality to make the product `in-stock` after the avail date has reached, then we need to maintain that functionality which would be outside the scope of our extension. I think this functionality is more suitable to be added to Woo core since it already allows a product to be back-ordered.

We could, however, only populate the avail date if the product stock status is set to backorder or preorder.

> Regarding the refactor for input and attributes, I'd say name and description could be useful in other locations, like exposing the attributes within the API. It just seems like more work to maintain an attribute by having to add/edit two classes.

Hmm. It just makes more sense in the current context to separate the concerns and define their inputs in another class. We could add the label and description back to the main attrib class if we ever need it.
